### DISCUSSION TITLE
error: Cannot dualize discrete problems

### DISCUSSION THREAD

**Discussion by 2xty** [yalmip/YALMIP#976] [Problems using YALMIP] [answered]
Hi guys, when I was learning Automatic dualization on Yalmip, it is said that "problems in YALMIP are modeled internally in the dual format (your problem is interpreted as a parametrization of the standard dual form). " 
But when I use the function 'dualize', it comes to error: Cannot dualize discrete problems. It bothers me that if Yalmip can't dualize discreat problems, how it sloves discreat problems?

appendix: the code in matlab
%------------------dual problem---------------------------
            cons3=[];
            cons3=[cons3,zpl(1,b1,t,ik)==0];
            cons3=[cons3,zpl(6,b1,t,ik)==0];
            cons3=[cons3,zpl(9,b1,t,ik)==0];
            for i=13:16
                cons3=[cons3,zpl(i,b1,t,ik)==0];
            end
                        
            for i=1:gi
                cons3=[cons3,pgk(i,b1,t,ik) <= A1(i,3) * xgist(i)];
                cons3=[cons3,pgk(i,b1,t,ik) >= A1(i,2) * xgist(i)];
            end
            
            for i=1:pl
                theta1 = thetak(:,b1,t,ik);
                bbm1 = ebbm(:,i);
                multi = theta1 .* bbm1;
                multi1 = sum(multi);
                
                cons3=[cons3,zpl(i,b1,t,ik)<=xplst(i)];

                cons3=[cons3,fplk(i,b1,t,ik) <= B1(i,5) * (xplst(i)-zpl(i,b1,t,ik))];
                cons3=[cons3,fplk(i,b1,t,ik) >= -1*B1(i,5) * (xplst(i)-zpl(i,b1,t,ik))];  

                cons3=[cons3,fplk(i,b1,t,ik) * B1(i,4) + multi1 <= BM * (1-xplst(i)+zpl(i,b1,t,ik))];
                cons3=[cons3,fplk(i,b1,t,ik) * B1(i,4) + multi1 >= -BM * (1-xplst(i)+zpl(i,b1,t,ik))];            
            end
            
            cons3=[cons3,ones(1,pl)*zpl(:,b1,t,ik)<=aplt];
           
            for i=1:b_1
                bbm1 = ebbm(i,:);
                fpl1 = fplk(:,b1,t,ik);
                multi = bbm1 * fpl1;
                bgm1 = ebgm(i,:);
                pg1 = pgk(:,b1,t,ik);
                multi1 = bgm1 * pg1;
                cons3=[cons3,cloadk(i,b1,t,ik)>=0];
                cons3=[cons3,multi + multi1 + cloadk(i,b1,t,ik) == eloadtb*C(i,2)];            
            end

            dobj = eblocktb*sum(cloadk(:,b1,t,ik));
            
            opts = sdpsettings('verbose',1,'solver','gurobi');
%             options.gurobi.MIPGAP=Gap;
            resu = optimize(cons3,dobj,sdpsettings('dualize',1));
![image](https://user-images.githubusercontent.com/82370096/131313203-dd320eef-88ef-4f27-a227-f2e6e036aab5.png)

**Comment by johanlofberg** [comment, marked as answer]
You simply cannot as there is no (agreed upon general) theory for duality when having discrete variables.

Why would you even be interested in that (what ever that would mean). You seem to have a very standard MILP

**Reply by 2xty** [reply]
Hi Johan, I'm excited to receive your letters. Here is the situation, I'm dealing with a standard MILP problem and need to dualize it to solve the C&CG subproblem. I used to get the dual form manually, but I found it complicated and error-prone when the primal form is complex. Thus I want to learn how to get the dual form of MILP problem automatically. 
And as there is no theory for duality when having discreate variables. How should I process the primal form to make it possible to get the dual form automatically? I believe there are ways...as problems in YALMIP are modeled internally in the dual format.  Looking forward to your reply!

**Reply by johanlofberg** [reply]
As I said, MILPs simply don't have a dual. Maybe you are talking about taking the continuous relaxation, and then working with the dual of that. Dualize will not be the tool though, it is used for completely different purposes (interpretation of symbolic form), and almost never relevant for LPs

Take your continuous relaxation, use export to some simple numerical form, and then the data defining the dual is trivial. Alternatively, use dualize on the continuous relaxation and then export that model.
